### PR TITLE
feat: use xcrun privacy instead of wix for location permission but keep using wix for rest

### PR DESCRIPTION
**Pull request by KazuCocoa**
For https://github.com/appium/appium/issues/18991

I wondered if we could move to `xcrun simctl privacy` since it looks like the Wix has an issue. What I could remember is maybe we had some issues in the wix command as old Xcode? but at least services listed in https://appium.github.io/appium-xcuitest-driver/5.10/execute-methods/#mobile-setpermission are in:

```
% xcrun simctl privacy
Grant, revoke, or reset privacy and permissions
Usage: simctl privacy <device> <action> <service> [<bundle identifier>]

	action
	     The action to take:
	         grant - Grant access without prompting. Requires bundle identifier.
	         revoke - Revoke access, denying all use of the service. Requires bundle identifier.
	         reset - Reset access, prompting on next use. Bundle identifier optional.
	     Some permission changes will terminate the application if running.
	service
	     The service:
	         all - Apply the action to all services.
	         calendar - Allow access to calendar.
	         contacts-limited - Allow access to basic contact info.
	         contacts - Allow access to full contact details.
	         location - Allow access to location services when app is in use.
	         location-always - Allow access to location services at all times.
	         photos-add - Allow adding photos to the photo library.
	         photos - Allow full access to the photo library.
	         media-library - Allow access to the media library.
	         microphone - Allow access to audio input.
	         motion - Allow access to motion and fitness data.
	         reminders - Allow access to reminders.
	         siri - Allow use of the app with Siri.
	bundle identifier
	     The bundle identifier of the target application.

Examples:
	reset all permissions: privacy <device> reset all
	grant test host photo permissions: privacy <device> grant photos com.example.app.test-host

Warning:
Normally applications must have valid Info.plist usage description keys and follow the API guidelines to request access to services. Using this command to bypass those requirements can mask bugs.

```

thus probably we could move to the native tools by Apple.

Some examples with this implementation:

```
[HTTP] --> POST /session/c011f8bd-b310-4fe4-bbfb-352f582e3f04/execute/sync
[HTTP] {"script":"mobile: setPermission","args":[{"bundleId":"com.kazucocoa.apple-samplecode.UICatalog","access":{"calendar":"yes","location-always":"yes","location":"yes"}}]}
[debug] [XCUITestDriver@a3fe (c011f8bd)] Calling AppiumDriver.execute() with args: ["mobile: setPermission",[{"bundleId":"com.kazucocoa.apple-samplecode.UICatalog","access":{"calendar":"yes","location-always":"yes","location":"yes"}}],"c011f8bd-b310-4fe4-bbfb-352f582e3f04"]
[debug] [XCUITestDriver@a3fe (c011f8bd)] Executing command 'execute'
[debug] [iOSSim] Setting access for 'com.kazucocoa.apple-samplecode.UICatalog': {
[debug] [iOSSim]   "calendar": "yes",
[debug] [iOSSim]   "location-always": "yes",
[debug] [iOSSim]   "location": "yes"
[debug] [iOSSim] }
[debug] [iOSSim] Granting 2 permissions for com.kazucocoa.apple-samplecode.UICatalog: location-always,location
[debug] [iOSSim] Setting permissions for com.kazucocoa.apple-samplecode.UICatalog wit applesimutils as {"calendar":"yes"}
[debug] [iOSSim] Executing: applesimutils --byId F9FF03FA-BAA2-4AC3-8BE2-0ACECEA054EB --bundle com.kazucocoa.apple-samplecode.UICatalog --setPermissions calendar\=yes
[debug] [iOSSim] Command output:
[debug] [XCUITestDriver@a3fe (c011f8bd)] Responding to client with driver.execute() result: null
[HTTP] <-- POST /session/c011f8bd-b310-4fe4-bbfb-352f582e3f04/execute/sync 200 436 ms - 14
```

Examples:
```
driver.execute_script 'mobile: setPermission', {bundleId: 'com.apple.Maps', access: {'location': 'yes', 'microphone': 'yes'}}
driver.execute_script 'mobile: setPermission', {bundleId: 'com.apple.Maps', access: {'location-always': 'yes', 'microphone': 'no'}}
driver.execute_script 'mobile: setPermission', {bundleId: 'com.apple.Maps', access: {'location-always': 'unset', 'microphone': 'unset'}}
driver.execute_script 'mobile: getPermission', {bundleId: 'com.apple.Maps', service: 'microphone'}
```

I'll create the documentation update as well in the xcuitest driver if it looks reasonable.

[update]
I have tested:

- Xcode 15 and iOS 17
- Xcode 14.3.1 and iOS 16.4

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Lets see how the test will be first...

**Comment by mykola-mokhnach** [comment]
I remember the `privacy` utilities were not working properly for iOS 15/16, so we had to use Wix instead. Perhaps, this has been fixed in v17: https://github.com/appium/appium-ios-simulator/pull/279

**Comment by KazuCocoa** [comment]
oh, i see.
I checked for ios 16.4 simulator with Xcode 14.3 right now with the same example command in the description. It seems to be working. Perhaps... it works with Xcode 14 and 15 env...?

File: test/functional/simulator-e2e-specs.js [hunk 1]
```diff
@@ -349,6 +349,17 @@ describe('advanced features', function () {
...
+
+  describe('Permission', function () {
+    it('should set and get', async function () {
+      await sim.setPermission('com.apple.Maps', 'microphone', 'yes');
+      await sim.getPermission('com.apple.Maps', 'microphone').should.eventually.eql('yes');
+      await sim.setPermission('com.apple.Maps', 'microphone', 'no');
+      await sim.getPermission('com.apple.Maps', 'microphone').should.eventually.eql('no');
+      await sim.setPermission('com.apple.Maps', 'microphone', 'unset');
+      await sim.getPermission('com.apple.Maps', 'microphone').should.eventually.eql('unset');
+    });
```
Comment (KazuCocoa) [inline test/functional/simulator-e2e-specs.js:361]: Below combination worked:
Functional Tests / test (13.1, iPhone X, 15)
Functional Tests / test (14.2, iPhone 14, 16.2)


Locally the same one worked with Xcode 15 and iOS 17.0, and Xcode 14.3 and iOS 16.4 simulators. So... in current our supported Xcode and iOS versions probably good with this...?

File: lib/extensions/permissions.js [hunk 2]
```diff
@@ -58,46 +53,54 @@ async function execSQLiteQuery (db, query) {
...
+      default:
+        log.warn(`${serviceName} does not support ${permissionsMapping[serviceName]}. Please specify yes, no or reset.`);
+    };
+  }
+
+  log.debug(`Granting permissions for ${bundleId} as ${JSON.stringify(grantPermissions)}`);
+  log.debug(`Revoking permissions for ${bundleId} as ${JSON.stringify(revokePermissions)}`);
+  log.debug(`Resetting permissions for ${bundleId} as ${JSON.stringify(resetPermissions)}`);
+
+  for (const action of grantPermissions) {
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:91]: can we perform commands in this loop in parallel?

File: lib/extensions/permissions.js [hunk 3]
```diff
@@ -58,46 +53,54 @@ async function execSQLiteQuery (db, query) {
...
+        break;
+      case STATUS.UNSET:
+        resetPermissions.push(serviceName);
+        break;
+      default:
+        log.warn(`${serviceName} does not support ${permissionsMapping[serviceName]}. Please specify yes, no or reset.`);
+    };
+  }
+
+  log.debug(`Granting permissions for ${bundleId} as ${JSON.stringify(grantPermissions)}`);
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:87]: Please only print the log if we have some permissions to change

File: test/functional/simulator-e2e-specs.js (hunk 1 above)
Comment (mykola-mokhnach) [reply test/functional/simulator-e2e-specs.js:361]: In the actual issue there was an issue with contacts permission. Have you checked that one as well?

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js:361]: Oh, contacts. Default apps do not require `contacts` permission. Let me create a thin app

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js:361]: You're correct, it seems like contacts did not work with code 15 x ios 17 as well. It looks like we should use wix only for contacts right now

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js:361]: Hm, but `applesimutils` also did not work for contacts

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js:361]: So, probably we could keep wix's one for below services

```
    camera=YES|NO|unset
    faceid=YES|NO|unset
    health=YES|NO|unset (iOS/tvOS 12.0 and above)
    homekit=YES|NO|unset
    location=always|inuse|never|unset
    notifications=YES|NO|critical|unset
    speech=YES|NO|unset
    userTracking=YES|NO|unset (iOS/tvOS 14.0 and above)
```

Comment (mykola-mokhnach) [reply test/functional/simulator-e2e-specs.js:361]: I believe we need to test Apple's privacy thing for all possible permission types. Also, as far as I remember, simctl forcefully restarts app under test. Not sure wix's utility requires that.

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js]: Actually Apple's one reset while the wix did not. Not sure if these process restart affects permission's behavior, but potentially we could use only `location` with the privacy method

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js]: Except for `contacts`, Apple's one worked but it reset the app process. So... probably we can use the apple's one only for `locations`

Comment (KazuCocoa) [reply test/functional/simulator-e2e-specs.js]: Updated. So, the latest implementation uses the privacy command only for the location stuff. Other commands by privacy works except for contacts, but the privacy command kills the app process, which might be breaking changes in terms of switching the method from wix.

File: lib/extensions/permissions.js [hunk 4]
```diff
@@ -12,6 +13,17 @@ const STATUS = Object.freeze({
 });
 
 const WIX_SIM_UTILS = 'applesimutils';
+
+// `location` permission does not work with WIX/applesimutils.
+// Note that except for 'contacts', the Apple's privacy command sets
+// permissions properly but it kills the app process while WIX/applesimutils does not.
+// In the backward compatibility perspective,
+// we'd like to keep the app process as possible.
+const APPLE_PRIVACY_CMD = [
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:22]: PERMISSIONS_APPLIED_VIA_SIMCTL

File: lib/extensions/permissions.js [hunk 5]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
+          grantPermissions.push(serviceName);
+          break;
+        case STATUS.NO:
+          revokePermissions.push(serviceName);
+          break;
+        case STATUS.UNSET:
+          resetPermissions.push(serviceName);
+          break;
+        default:
+          log.warn(`${serviceName} does not support ${permissionsMapping[serviceName]}. Please specify yes, no or unset.`);
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:130]: I assume we should throw an exception here

File: lib/extensions/permissions.js [hunk 6]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
+      };
+    }
+  }
+
+  let permissionPromises = [];
+
+  if (!_.isEmpty(grantPermissions)) {
+    log.debug(`Granting permissions for ${bundleId} as ${JSON.stringify(grantPermissions)}`);
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:138]:
```
Granting ${util.pluralize(grantPermissions.length, 'permission', false)}: ${grantPermissions}
```

same below

File: lib/extensions/permissions.js [hunk 7]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
+
+  if (!_.isEmpty(resetPermissions)) {
+    log.debug(`Revoking permissions for ${bundleId} as ${JSON.stringify(resetPermissions)}`);
+    for (const action of resetPermissions) {
+      permissionPromises.push(simctl.resetPermission(bundleId, action));
+    }
+  }
+
+  if (!_.isEmpty(permissionPromises)) {
+    B.all(permissionPromises);
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:159]: await

File: lib/extensions/permissions.js [hunk 8]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
-    .join(',');
-  return await execWix([
-    '--byId', udid,
-    '--bundle', bundleId,
-    '--setPermissions', permissionsArg,
-  ]);
+async function setAccess (simctl, udid, bundleId, permissionsMapping) {
+  const wixPermissions = {};
+
+  const grantPermissions = [];
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:111]: `/** @type {string[]} */`

File: test/functional/simulator-e2e-specs.js [hunk 9]
```diff
@@ -349,6 +349,28 @@ describe('advanced features', function () {
       });
     });
   });
+
+  describe('Permission', function () {
+    it('should set and get with simctrl privacy command', async function () {
+      // no exceptions
+      expect(await sim.setPermission('com.apple.Maps', 'location', 'yes')).should.not.rejected;
```
Comment (mykola-mokhnach) [inline test/functional/simulator-e2e-specs.js:356]: should.not.be.rejected

Comment (mykola-mokhnach) [reply test/functional/simulator-e2e-specs.js:356]: or `should.be.fulfilled`

File: lib/extensions/permissions.js [hunk 10]
```diff
@@ -12,6 +13,17 @@ const STATUS = Object.freeze({
 });
 
 const WIX_SIM_UTILS = 'applesimutils';
+
+// `location` permission does not work with WIX/applesimutils.
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:17]: perhaps, we should also report this as an issue to Wix?

File: lib/extensions/permissions.js [hunk 11]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
+        case STATUS.UNSET:
+          resetPermissions.push(serviceName);
+          break;
+        default:
+          log.errorAndThrow(`${serviceName} does not support ${permissionsMapping[serviceName]}. Please specify 'yes', 'no' or 'unset'.`);
+      };
+    }
+  }
+
+  let permissionPromises = [];
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:135]: you could make this array typed as well

File: lib/extensions/permissions.js [hunk 12]
```diff
@@ -82,22 +94,84 @@ async function execWix (args) {
...
-  const permissionsArg = _.toPairs(permissionsMapping)
-    .map((x) => `${x[0]}=${formatStatus(x[1])}`)
-    .join(',');
-  return await execWix([
-    '--byId', udid,
-    '--bundle', bundleId,
-    '--setPermissions', permissionsArg,
-  ]);
+async function setAccess (simctl, udid, bundleId, permissionsMapping) {
+  const wixPermissions = {};
```
Comment (mykola-mokhnach) [inline lib/extensions/permissions.js:109]: this object could be typed

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Some tests are unhappy though

File: lib/extensions/permissions.js (hunk 10 above)
Comment (KazuCocoa) [reply lib/extensions/permissions.js:17]: Found https://github.com/wix/AppleSimulatorUtils/issues/111 in their repo. Let me link here

File: lib/extensions/permissions.js (hunk 12 above)
Comment (KazuCocoa) [reply lib/extensions/permissions.js:109]: oh. record. thanks

**Comment by KazuCocoa** [comment]
Ok, the tests are finally happy now

**Comment by fleytman** [comment]
`location-always does not support YES. Please specify 'yes', 'no' or 'unset'.`
Thank you for this feature! Alas, it's not very obvious to write for some permissions in upper case and others in lower case. I think you should make it upper case for location too, or make it case-insensitive.

**Comment by KazuCocoa** [comment]
5.11.8 should be ok. It handles the input for the location as case insensitive.

**Pull request merged**